docs/design/run.md

```diff
@@ -0,0 +1,275 @@
+immediately error out, as published history should be immutable. We may want to
+support a `--force` flag for an override but it won't be available in the first
+iteration of the command. 
+
+### Private/Draft revisions:
+
+For private/draft revisions, we just amend the changes, as Jujutsu usually does. 
+We also expose the actual behavior as a command option.
+
+## Execution order/parallelism:
+
+It may be useful to execute commands in topological order. For example, 
+commands with costs proportional to incremental changes, like build systems. 
+There may also be other revelant heuristics, but topological order is an easy
+and effective way to start. 
+
+Parallel execution of commands on different commits may choose to schedule 
+commits to still reduce incremental changes in the working copy used by each
+execution slot/"thread". However, running the command on all commits 
+concurrently should be possible if desired. 
+
+Executing commands in topological order allows for more meaningful use of any 
+potential features that stop execution "at the first failure". For example, 
+when running tests on a chain of commits, it might be useful to proceed in 
+topological/chronological order, and stop on the first failure, because it 
+might imply that the remaining executions will be undesirable because they will
+also fail.
+
+## Dealing with failure:
+
+It will be useful to have multiple strategies to deal with failures on a single
+or multiple revisions. The reason for these strategies is to allow customized
+conflict handling. These strategies then can be exposed in the ui with a 
+matching command.
+
+**Continue:** If any subprocess fails, we will continue the work on child 
+revisions. Notify the user on exit about the failed revisions. 
+
+**Stop:** Signal a fatal failure and cancel any scheduled work that has not
+yet started running, but let any already started subprocess finish. Notify the
+user about the failed command and display the generated error from the 
+subprocess. 
+
+**Fatal:** Signal a fatal failure and immediately stop processing and kill any 
+running processes. Notify the user that we failed to apply the command to the 
+specific revision. 
+
+We will leave any affected in its current state, if any subprocess fails. This
+allows us provide a better user experience, as leaving revisions in an 
+undesirable state, e.g partially formatted, may confuse users.
+
+## Resource constraints:
+
+It will be useful to constrain the execution to prevent resource exhaustion. 
+Relevant resources could include:
+- CPU and memory available on the machine running the commands. `jj run` can
+provide some simple mitigations like limiting parallelism to "number of CPUs" 
+by default, and limiting parallelism by dividing "available memory" by some 
+estimate or measurement of per-invocation memory use of the commands.
+- External resourec that are not immediately known to jj. For example, 
```

Comment:
typo in "resource" (btw, you could run a spell-checker on the whole doc)

Reply:
corrected